File: csharp/BronKerbosch/BronKerbosch2aGP.cs
// Bron-Kerbosch algorithm with pivot of highest degree (IK_GP)

using BronKerbosch;
using System.Collections.Generic;
using System.Linq;

internal static class BronKerbosch2aGP<VertexSet, VertexSetMgr, TAccumulator>
    where VertexSet : ISet<Vertex>
    where VertexSetMgr : IVertexSetMgr<VertexSet>
    where TAccumulator : ICliqueAccumulator<TAccumulator>, new()
{
    public static void Explore(UndirectedGraph<VertexSet, VertexSetMgr> graph,
                               CliqueConsumer<TAccumulator> consumer)
    {
        VertexSet candidates = VertexSetMgr.From(graph.ConnectedVertices());
        if (candidates.Any())
        {
            Pivot<VertexSet, VertexSetMgr, TAccumulator>.Visit(
                graph,
                consumer,
                PivotChoice.MaxDegreeLocal,
                candidates,
                VertexSetMgr.Empty(),
                []);
        }
    }
}

File: go/BronKerbosch/bron_kerbosch2a_gp.go
package BronKerbosch

func bronKerbosch2aGP(graph *UndirectedGraph, consumer Consumer) {
	// Bron-Kerbosch algorithm with pivot of highest degree within remaining candidates
	// chosen from candidates only (IK_GP).
	candidates := graph.connectedVertices()
	excluded := make(VertexSet, len(candidates))
	visit(graph, consumer, MaxDegreeLocal, candidates, excluded, nil)
	consumer.close()
}
